src/mul.nr

```diff
@@ -0,0 +1,272 @@
+use crate::utils::arrayX::ArrayX;
+use crate::BigNum;
+use crate::bignum_test::{Fq, Fqq};
+use crate::fields::bn254Fq::BNParams as BNParams;
+use crate::BigNumParamsTrait;
+use crate::test_params::{
+  nonprime3000::Test3000NonPrimeParams,
+  prime256::Test256PrimeParams,
+  };
+
+/**
+  * @brief schoolbook multiplication of 2 N-length arrays, no reductions
+  **/
+pub fn schoolbook_mult<N>(a: [Field; N], b: [Field; N]) -> ArrayX<Field, N, 2> {
+    let mut res = ArrayX { segments: [[0;N], [0; N]] };
+    for i in 0..N {
+        for j in 0..N {
+            res.add_assign(i+j, a[i] * b[j]);
+        }
+    }
+    res
+}
+
+/**
+  * @brief simple addition of limbs of 2 N-length arrays, no reductions
+  **/
+fn add_arrays<N>(a: [Field; N], b: [Field; N]) -> [Field;N] {
+    let mut res = [0;N];
+    for i in 0..N {
+        res[i] = a[i]+b[i];
+    }
+    res
+}
+
+/**
+  * @brief simple subtraction of limbs of 2 N-length arrays, no reductions
+  * @description This directly subtracts a[i]-b[i]-c[i] for all N limbs,
+  *              thus some limbs might have a "negative" value which has wrapped
+  *              around in the Field type.
+  **/
+fn subtract_arrays<N>(a: ArrayX<Field, N, 2>, b: ArrayX<Field, N, 2>, c: ArrayX<Field, N, 2>) -> ArrayX<Field, N, 2> {
+    let mut res = ArrayX { segments: [[0;N], [0; N]] };
+    for i in 0..2 {
+        for j in 0..N {
+            res.segments[i][j] = a.segments[i][j] - b.segments[i][j] - c.segments[i][j]; 
+        }
+    }
+    res
+}
```

Comment:
this is actually a possible improvement the other way around I was thinking about. I think accessing `segments[i][j]` might be faster than `segments.get(i)`, because the latter calculates the value of `i` and `j` which uses a division and a mod. So I think we can try to replace and if it makes any difference. I agree with you that if they have the same gatecount, `get()` and `set()` is more readable.